mne/beamformer/_lcmv.py

```diff
@@ -67,28 +67,46 @@ def _setup_picks(picks, info, forward, noise_cov=None):
 @verbose
 def _apply_lcmv(data, info, tmin, forward, noise_cov, data_cov, reg,
                 label=None, picks=None, pick_ori=None, rank=None,
+                noise_norm = False, whitening=False, weight_norm=None, 
                 verbose=None):
     """LCMV beamformer for evoked data, single epochs, and raw data."""
     is_free_ori, ch_names, proj, vertno, G = \
         _prepare_beamformer_input(info, forward, label, picks, pick_ori)
 
-    # Handle whitening + data covariance
-    whitener, _ = compute_whitener(noise_cov, info, picks, rank=rank)
-
-    # whiten the leadfield
-    G = np.dot(whitener, G)
-
-    # Apply SSPs + whitener to data covariance
+    # data covariance
     data_cov = pick_channels_cov(data_cov, include=ch_names)
     Cm = data_cov['data']
-    if info['projs']:
-        Cm = np.dot(proj, np.dot(Cm, proj.T))
-    Cm = np.dot(whitener, np.dot(Cm, whitener.T))
 
-    # Tikhonov regularization using reg parameter to control for
+    if whitening:
+        # Handle whitening + data covariance
+        whitener, _ = compute_whitener(noise_cov, info, picks, rank=rank)
+    
+        # whiten the leadfield
+        G = np.dot(whitener, G)
+    
+        # Apply SSPs + whitener to data covariance
+        if info['projs']:
+            Cm = np.dot(proj, np.dot(Cm, proj.T))
+        Cm = np.dot(whitener, np.dot(Cm, whitener.T))
+
+    # Tikhonov regularization using reg parameter d to control for
     # trade-off between spatial resolution and noise sensitivity
-    Cm_inv = _reg_pinv(Cm.copy(), reg)
+    Cm_inv, d = _reg_pinv(Cm.copy(), reg)
+
+    # estimate noise level based on covariance matrix, taking the smallest
+    # eigenvalue that is not zero       
+    _, noise, _ = np.linalg.svd(Cm)
+    noise = noise[np.linalg.matrix_rank(Cm)-1]
+    
+    # use either noise floor or regularization parameter d
+    noise = np.max([noise, d])           
     del Cm
+    
+    # Compute square of Cm_inv used for weight normalization
+    Cm_inv_sq = np.linalg.matrix_power(Cm_inv,2)
```

Comment:
and keep in mind the whitespaces .... sounds like a nitpick but it's a strong convention